include/maliput_sparse/builder/builder.h

```diff
@@ -0,0 +1,219 @@
+#pragma once
+/**
+ * @file Builder API to construct a maliput::api::RoadGeometry.
+ *
+ * @details It allows simple construction of nested nodes in the maliput::api::RoadGeometry graph. The
+ * geometry details are expected to be computed / loaded outside this method to allow parallelization
+ * of that process. The graph construction is a synchronous operation though and requires
+ * that the geometry information is available when building maliput::api::Lanes.
+ *
+ * A super simple 2-lane dragway could be constructed as follows:
+ *
+ * @code{cpp}
+ * const maliput::math:Vector3 start_left_lane_1{0., 0., 0.};
+ * const maliput::math:Vector3 end_left_lane_1{100., 0., 0.};
+ * const maliput::math:Vector3 start_right_lane_1{0., 5., 0.};
+ * const maliput::math:Vector3 end_right_lane_1{100., 5., 0.};
+ * const maliput::math:Vector3 start_left_lane_2 = start_right_lane_1;
+ * const maliput::math:Vector3 end_left_lane_2 = end_right_lane_1;
+ * const maliput::math:Vector3 start_right_lane_2{0., 10., 0.};
+ * const maliput::math:Vector3 end_right_lane_2{100., 10., 0.};
+ *
+ * LineString3d left_line_string_1{start_left_lane_1, end_left_lane_1};
+ * LineString3d right_line_string_1{start_right_lane_1, end_right_lane_1};
+ * LineString3d left_line_string_2{start_left_lane_2, end_left_lane_2};
+ * LineString3d right_line_string_2{start_right_lane_2, end_right_lane_2};
+ *
+ * RoadGeometryBuilder rg_builder;
+ * std::unique_ptr<RoadGeometry> road_geometry = RoadGeometryBuilder()
+ *   .Id("two_lane_dragway")
+ *   .StartJunction()
+ *     .Id("j0")
+ *     .StartSegment()
+ *       .Id("j0_s0")
+ *       .StartLane()
+ *         .Id("j0_s0_l1")
+ *         .StartLaneGeometry()
+ *           .LeftLineString(left_line_string_1)
+ *           .RightLineString(right_line_string_1)
+ *         .EndLaneGeometry()
+ *       .EndLane()
+ *       .StartLane()
+ *         .Id("j0_s0_l2")
+ *         .StartLaneGeometry()
+ *           .LeftLineString(left_line_string_2)
+ *           .RightLineString(right_line_string_2)
+ *         .EndLaneGeometry()
+ *       .EndLane()
+ *     .EndSegment()
+ *   .EndJunction()
+ *   .Build();
+ * @endcode{cpp}
+ */
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <maliput/api/junction.h>
+#include <maliput/api/lane.h>
+#include <maliput/api/road_geometry.h>
+#include <maliput/api/segment.h>
+#include <maliput/common/passkey.h>
+#include <maliput/geometry_base/junction.h>
+#include <maliput/geometry_base/lane.h>
+#include <maliput/geometry_base/road_geometry.h>
+#include <maliput/geometry_base/segment.h>
+
+#include "maliput_sparse/geometry/line_string.h"
+
+namespace maliput_sparse {
+namespace builder {
+namespace details {
+
+class BuilderBase {
+ public:
+  virtual ~BuilderBase() = default;
+};
+
+template <typename ParentT>
+class NestedBuilder : public BuilderBase {
+ public:
+  explicit NestedBuilder(ParentT* parent) : parent_(parent) { MALIPUT_THROW_UNLESS(parent_ != nullptr); }
```

Comment:
I find this nested builder convenient and natural to use at some point but I have some concerns when for example thinking downstream for the backends.

In order to go down in the hierarchy this api requires for me to know, pretty decently, how is the road geometry composed in a maliput-like fashion. Namely, to add a lane I need to know in which segment I will add it and in which junction I will add the segment.

From a user perspective I am not that confident that the user will be able to know this when parsing a file with a X format for road describing being loaded. Maybe, being able to load/build a lane and later on deciding which segment it should be added to, could be necessary for this situation? Wdyt @agalbachicar ?

Reply:
I think that the scope of this API should be to only build a RoadGeometry. Any translation between the specific map format and the maliput graph must be implemented in the parser. With that responsibility separation in mind, we can change the API of the Builder to provider certain guarantees about what we will do with the provided information or favor one use or the other, but in the end, I think the closer to the RoadGeometry graph the better.

One thing that is a bit obscure and perhaps not too obvious is how BranchPoints will be created. I plan to work a bit more on that later by enabling a policy in the RoadGeometryBuilder to create BranchPoints:
- by LaneEnd overlap / intersection.
- by manually creating LaneEnds and LaneEndSets.

This is another relevant aspect of the connecitivity graph that is not contemplated for the moment.

Reply:
> Any translation between the specific map format and the maliput graph must be implemented in the parser

SG.

> One thing that is a bit obscure and perhaps not too obvious is how BranchPoints will be created.

>- by LaneEnd overlap / intersection.

For this non-manual case: Lanelet2 solves the connectivity by taking a look at the points that are shared among lanes. Two points defining an end of a lane and a start on another lane define a connection. We haven't defined having this kind of sharable data (points/linesstrings) but it is worth mentioning.

Reply:
> For this non-manual case: Lanelet2 solves the connectivity by taking a look at the points that are shared among lanes. Two points defining an end of a lane and a start on another lane define a connection. We haven't defined having this kind of sharable data (points/linesstrings) but it is worth mentioning.

This is an interesting idea... I guess we could have some sort of associative container to index points and lanes. The key thing here is that so far we always linked centerline from Lane X with centerline from Lane Y. We might need to expose that information in the LaneGeometry to accomplish that before everything is created.